R/unnecessary_nesting_linter.R

```diff
@@ -0,0 +1,168 @@
+#' Block instances of excessive nesting
+#'
+#' Excessive nesting harms readability. Use helper functions or early returns
+#'   to reduce nesting wherever possible.
+#'
+#' @examples
+#' # will produce lints
+#' code <- "if (A) {\n  stop('A is bad!')\n} else {\n  do_good()\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = unnecessary_nesting_linter()
+#' )
+#'
+#' code <- "tryCatch(\n  {\n    foo()\n  },\n  error = identity\n)"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = unnecessary_nesting_linter()
+#' )
+#'
+#' # okay
+#' code <- "if (A) {\n  stop('A is bad because a.')\n} else {\n  stop('!A is bad too.')\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = unnecessary_nesting_linter()
+#' )
+#'
+#' code <- "capture.output({\n  foo()\n})"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = unnecessary_nesting_linter()
+#' )
+#'
+#' @evalRd rd_tags("unnecessary_nesting_linter")
+#' @seealso
+#'  - [cyclocomp_linter()] for another linter that penalizes overly complexcode.
+#'  - [linters] for a complete list of linters available in lintr.
+#' @export
+unnecessary_nesting_linter <- function() {
+  exit_calls <- c("stop", "return", "abort", "quit", "q")
+  # These calls can be called in the sibling branch and not trigger a lint,
+  #   allowing for cleanly parallel code, where breaking it would often harm readability:
+  #   if (A) {
+  #     stop()
+  #   } else {
+  #     warning()
+  #   }
+  # NB: print() is intentionally excluded since its usage is usually a mistake (?print_linter)
+  signal_calls <- c(
+    exit_calls,
+    "warning", "warn", "message", "cat", "LOG", "stopifnot"
```

Comment:
yep... left it in here as a reminder to discuss what sort of parameterization we should build in here, WDYT?

`allow_signal_calls = c("warning", "message", "cat", "stopifnot")`?

Reply:
I'm not 100% sure yet.

Reply:
So, just to understand this: These functions explicitly allow a parallel branch in `if` statements with a single control-flow breaking statement?

Reply:
Right... they're allowed to pair with "exit" clauses, but can't cause a lint by themselves -- nested code using them can't _always_ be unnested. IOW they can remove a lint, but not cause a lint.

Reply:
i.e. this won't lint:

```r
if (A) {
  warning('hi')
} else {
  do()
}
```

Nor will this:

```r
if (A) {
  warning('hi')
} else {
  stop('bye')
}
```

But this will:

```r
if (A) {
  stop('hi')
} else {
  do()
}
```

Reply:
What about

```r
if (A) {
  stop()
} else {
  do()
  warning()
}
```

Reply:
won't lint -- only the terminal call is examined

Reply:
So
```r
if (A) {
  stop()
} else {
  warning()
  do()
}
```
Would lint?

Reply:
yes

Reply:
Of course, one could always disable this behavior (and always force unnesting if possibe) by passing e.g. `character()` to the argument accepting a vector of calls.

Reply:
I'm just wondering how intuitive the exception logic is.